### PR TITLE
bpo-40389: Improve repr of typing.Optional

### PR DESCRIPTION
**Pull request by Endilll**
https://bugs.python.org/issue40389

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@Endilll

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/typing.py [hunk 1]
```diff
@@ -683,6 +683,9 @@ def copy_with(self, params):
         return _GenericAlias(self.__origin__, params, name=self._name, inst=self._inst)
 
     def __repr__(self):
+        if (self.__origin__ == Union and len(self.__args__) == 2
+                and self.__args__[1] is type(None)):
```
Comment (remilapeyre) [inline Lib/typing.py]: Hi @Endilll, thanks for opening a pull request for this. Notice that this does not work, when `None` is the first element:
```
>>> import typing
>>> typing.Union[None, int]
typing.Union[NoneType, int]
>>> typing.Union[None, int] == typing.Optional[int]
True
```

This change will also need some tests and an NEWS entry.

**Review by remilapeyre — CHANGES_REQUESTED** [review]

**Comment by remilapeyre** [comment]
Ya NEWS entry using https://blurb-it.herokuapp.com/

File: Lib/typing.py (hunk 1 above)
Comment (Endilll) [reply Lib/typing.py]: Thank you for the feedback. Fixed.

**Comment by aeros** [comment]
Thanks for the PR and welcome @Endilll. 

FYI: assuming the idea is approved by the maintainers of `typing`, this change will most definitely require some updates to the tests in `test_typing.py` (in `Lib/test`).

**Review by ilevkivskyi — APPROVED** [review]
Looks good, thanks! Also be sure to sign the CLA.

The Azure failure look like a flake, but I can't re-trigger it, and it is required for merging. @zooba do you know how to fix this?

**Review by gvanrossum — APPROVED** [review]
LGTM. Please sign the CLA so we can merge.

**Comment by Endilll** [comment]
I've submitted CLA back on Saturday, and bot says that one business day needed to update the records, so I suppose we just need to wait a bit.

**Comment by gvanrossum** [comment]
Closing and reopening to re-run the Azure pipelines.

**Comment by gvanrossum** [comment]
Congratulations!

**Comment by Endilll** [comment]
I want to thank everybody for such a welcoming attitude.

**Pull request merged**